docker, containerd: image registry mirrors

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Resolves https://github.com/bottlerocket-os/bottlerocket/issues/1577


**Description of changes:**
```
Author: Erikson Tung <etung@amazon.com>
Date:   Wed Jun 23 15:22:11 2021 -0700

    migrations: migrate `settings.container-registry`, and related models
    
    Adds migrations for the new `container-registry` setting and the new
    docker daemon config file and the new config file for host-containers
    and bootstrap containers.

```
```
Author: Erikson Tung <etung@amazon.com>
Date:   Mon Jul 26 22:52:09 2021 -0700

    host-ctr: add support for container image registries
    
    Adds support for configuring container image registrys for
    host-containers and bootstrap containers. A lot of the logic is
    borrowed from containerd-cri plugin's implementation of image registry
    mirrors.

```
```
Author: Erikson Tung <etung@amazon.com>
Date:   Tue Jun 22 15:12:14 2021 -0700

    docker, containerd: container image registry mirrors
    
    Adds a new setting `settings.container-registry.mirrors` that lets users set
    pull-through caches for container image registries.

```

**Testing done:**

Built AMIs and launched an ECS host and an K8S host.

Set up a pull-through cache for docker hub on a separate host:
```
sudo docker run \
  -d \
  -p 80:5000 \
  --restart=always \
  --name=through-cache \
  -e REGISTRY_PROXY_REMOTEURL="https://registry-1.docker.io" \
  -e REGISTRY_PROXY_USERNAME=snip \
  -e REGISTRY_PROXY_PASSWORD=snip \
  registry
```

Set the image registry mirror to the pull-through cache host on both ECS and K8S hosts:
```
apiclient set --json '{"container-registry":{"mirrors":{"docker.io":["http://my-pull-through-cache-host-ip"]}}}'
```

Tried running a nginx task on in the ECS cluster, observed that the image was pulled from the cache
The pull-through cache registry is responding to the requests for manifests and blobs with HTTP status code 200:
```
time="2021-06-23T20:40:09.738518303Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=75a15a82-6b80-4508-ad7a-9cc17b1a1e84 http.request.method=GET http.request.remoteaddr="54.184.78.250:44052" http.request.uri="/v2/" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/json; charset=utf-8" http.response.duration=1.016725ms http.response.status=200 http.response.written=2 
54.184.78.250 - - [23/Jun/2021:20:40:09 +0000] "GET /v2/ HTTP/1.1" 200 2 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.494864674Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=de6ec364-34d4-4648-8f30-9613f5f23512 http.request.method=HEAD http.request.remoteaddr="54.184.78.250:44054" http.request.uri="/v2/library/nginx/manifests/mainline" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/vnd.docker.distribution.manifest.list.v2+json" http.response.duration=753.789996ms http.response.status=200 http.response.written=1862 
54.184.78.250 - - [23/Jun/2021:20:40:09 +0000] "HEAD /v2/library/nginx/manifests/mainline HTTP/1.1" 200 1862 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/manifests/sha256:47ae43cdfc7064d28800bc42e79a429540c7c80168e8c8952778c0d5af1c09db HTTP/1.1" 200 1862 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.509265258Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=b60c65c4-6909-4b6a-89dc-a5e95add0ec8 http.request.method=GET http.request.remoteaddr="54.184.78.250:44056" http.request.uri="/v2/library/nginx/manifests/sha256:47ae43cdfc7064d28800bc42e79a429540c7c80168e8c8952778c0d5af1c09db" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/vnd.docker.distribution.manifest.list.v2+json" http.response.duration=1.813035ms http.response.status=200 http.response.written=1862 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/manifests/sha256:2f1cd90e00fe2c991e18272bb35d6a8258eeb27785d121aa4cc1ae4235167cfd HTTP/1.1" 200 1570 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.525015041Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=da96c670-692a-49aa-8eee-a20a3cf115cb http.request.method=GET http.request.remoteaddr="54.184.78.250:44058" http.request.uri="/v2/library/nginx/manifests/sha256:2f1cd90e00fe2c991e18272bb35d6a8258eeb27785d121aa4cc1ae4235167cfd" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.response.duration=1.788932ms http.response.status=200 http.response.written=1570 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:b82f7f888feb03d38fed4dad68d7265a8b276f1f0c543d549fc6ef30b42c00eb HTTP/1.1" 200 667 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.550816892Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=83e13e41-e792-4393-a72e-5dba761a479f http.request.method=GET http.request.remoteaddr="54.184.78.250:44072" http.request.uri="/v2/library/nginx/blobs/sha256:b82f7f888feb03d38fed4dad68d7265a8b276f1f0c543d549fc6ef30b42c00eb" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=5.145482ms http.response.status=200 http.response.written=667 
time="2021-06-23T20:40:10.553786666Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=844055df-4f8b-4f52-bbb6-cf76cd6e1699 http.request.method=GET http.request.remoteaddr="54.184.78.250:44066" http.request.uri="/v2/library/nginx/blobs/sha256:03e6a245275128e26fc650e724e3fc4510d81f8111bae35ece70242b0a638215" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=5.994485ms http.response.status=200 http.response.written=895 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:03e6a245275128e26fc650e724e3fc4510d81f8111bae35ece70242b0a638215 HTTP/1.1" 200 895 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.55698674Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=9b3cea70-112a-43ea-bcca-36e4f4107980 http.request.method=GET http.request.remoteaddr="54.184.78.250:44060" http.request.uri="/v2/library/nginx/blobs/sha256:4f380adfc10f4cd34f775ae57a17d2835385efd5251d6dfe0f246b0018fb0399" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=23.66608ms http.response.status=200 http.response.written=7733 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:4f380adfc10f4cd34f775ae57a17d2835385efd5251d6dfe0f246b0018fb0399 HTTP/1.1" 200 7733 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.560957612Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=2c20e6a6-e5e3-4d06-bd58-ef49fd6572ef http.request.method=GET http.request.remoteaddr="54.184.78.250:44068" http.request.uri="/v2/library/nginx/blobs/sha256:b21fed559b9f420d83f8e38ca08d1ac4f15298a3ae02c6de56f364bee2299f78" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=7.948473ms http.response.status=200 http.response.written=602 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:b21fed559b9f420d83f8e38ca08d1ac4f15298a3ae02c6de56f364bee2299f78 HTTP/1.1" 200 602 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.56643432Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=9bb0785e-9832-4ecf-89df-db7328c4f5ef http.request.method=GET http.request.remoteaddr="54.184.78.250:44070" http.request.uri="/v2/library/nginx/blobs/sha256:5430e98eba646ef4a34baff035f6f7483761c873711febd48fbcca38d7890c1e" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=4.082108ms http.response.status=200 http.response.written=1397 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:5430e98eba646ef4a34baff035f6f7483761c873711febd48fbcca38d7890c1e HTTP/1.1" 200 1397 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:b4d181a07f8025e00e0cb28f1cc14613da2ce26450b80c54aea537fa93cf3bda HTTP/1.1" 200 27145851 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.754292088Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=63d7ffbc-1b7d-47cb-8357-54672a3de54f http.request.method=GET http.request.remoteaddr="54.184.78.250:44062" http.request.uri="/v2/library/nginx/blobs/sha256:b4d181a07f8025e00e0cb28f1cc14613da2ce26450b80c54aea537fa93cf3bda" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=220.138344ms http.response.status=200 http.response.written=27145851 
54.184.78.250 - - [23/Jun/2021:20:40:10 +0000] "GET /v2/library/nginx/blobs/sha256:edb81c9bc1f5416a41e5bea21748dc912772fedbd4bd90e5e3ebfe16b453edce HTTP/1.1" 200 26580118 "" "docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)"
time="2021-06-23T20:40:10.767215744Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=cdbbbb46-0c3e-42da-b1db-18717b747376 http.request.method=GET http.request.remoteaddr="54.184.78.250:44064" http.request.uri="/v2/library/nginx/blobs/sha256:edb81c9bc1f5416a41e5bea21748dc912772fedbd4bd90e5e3ebfe16b453edce" http.request.useragent="docker/20.10.4 go/go1.16.3 git-commit/363e9a88a11be517d9e8c65c998ff56f774eb4dc kernel/5.10.35 os/linux arch/amd64 UpstreamClient(Go-http-client/1.1)" http.response.contenttype="application/octet-stream" http.response.duration=233.865597ms http.response.status=200 http.response.written=26580118 

```

Ran an nginx pod in my K8S cluster that has a single node:
```
kubectl --kubeconfig bottlerocket.kubeconfig run nginx --image=nginx
```
The image got pulled from the pull-through cache registry just like above:
```
18.236.180.8 - - [23/Jun/2021:19:36:37 +0000] "HEAD /v2/library/nginx/manifests/latest?ns=docker.io HTTP/1.1" 200 1862 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:37.944166513Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=a2678ced-2d46-4fe3-9595-56d04ab029c4 http.request.method=HEAD http.request.remoteaddr="18.236.180.8:50412" http.request.uri="/v2/library/nginx/manifests/latest?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/vnd.docker.distribution.manifest.list.v2+json" http.response.duration=728.409568ms http.response.status=200 http.response.written=1862 
18.236.180.8 - - [23/Jun/2021:19:36:37 +0000] "GET /v2/library/nginx/manifests/sha256:47ae43cdfc7064d28800bc42e79a429540c7c80168e8c8952778c0d5af1c09db?ns=docker.io HTTP/1.1" 200 1862 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:37.960599554Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=abbec0a6-6372-4aef-9ef3-d0c71d9c8a9b http.request.method=GET http.request.remoteaddr="18.236.180.8:23923" http.request.uri="/v2/library/nginx/manifests/sha256:47ae43cdfc7064d28800bc42e79a429540c7c80168e8c8952778c0d5af1c09db?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/vnd.docker.distribution.manifest.list.v2+json" http.response.duration=4.318429ms http.response.status=200 http.response.written=1862 
time="2021-06-23T19:36:38.220117889Z" level=info msg="Adding new scheduler entry for library/nginx@sha256:2f1cd90e00fe2c991e18272bb35d6a8258eeb27785d121aa4cc1ae4235167cfd with ttl=167h59m59.999997602s" go.version=go1.11.2 instance.id=1470de18-b2bd-4f24-b811-4f48dd848efb service=registry version=v2.7.1 
18.236.180.8 - - [23/Jun/2021:19:36:37 +0000] "GET /v2/library/nginx/manifests/sha256:2f1cd90e00fe2c991e18272bb35d6a8258eeb27785d121aa4cc1ae4235167cfd?ns=docker.io HTTP/1.1" 200 1570 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:38.220422328Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=e4effd51-2b0f-4092-878b-c6c744d2815c http.request.method=GET http.request.remoteaddr="18.236.180.8:23923" http.request.uri="/v2/library/nginx/manifests/sha256:2f1cd90e00fe2c991e18272bb35d6a8258eeb27785d121aa4cc1ae4235167cfd?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.response.duration=242.563687ms http.response.status=200 http.response.written=1570 
time="2021-06-23T19:36:38.57423413Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=207b8564-62f8-4ddd-941d-0af91e6d30d3 http.request.method=GET http.request.remoteaddr="18.236.180.8:23923" http.request.uri="/v2/library/nginx/blobs/sha256:4f380adfc10f4cd34f775ae57a17d2835385efd5251d6dfe0f246b0018fb0399?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=335.486858ms http.response.status=200 http.response.written=7733 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:4f380adfc10f4cd34f775ae57a17d2835385efd5251d6dfe0f246b0018fb0399?ns=docker.io HTTP/1.1" 200 7733 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:38.659022282Z" level=info msg="Adding new scheduler entry for library/nginx@sha256:4f380adfc10f4cd34f775ae57a17d2835385efd5251d6dfe0f246b0018fb0399 with ttl=167h59m59.999998257s" go.version=go1.11.2 instance.id=1470de18-b2bd-4f24-b811-4f48dd848efb service=registry version=v2.7.1 
time="2021-06-23T19:36:38.956560519Z" level=info msg="Adding new scheduler entry for library/nginx@sha256:5430e98eba646ef4a34baff035f6f7483761c873711febd48fbcca38d7890c1e with ttl=167h59m59.999997352s" go.version=go1.11.2 instance.id=1470de18-b2bd-4f24-b811-4f48dd848efb service=registry version=v2.7.1 
time="2021-06-23T19:36:38.976072094Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=be13ba01-f5ab-4f85-ae70-d7e6225b55f8 http.request.method=GET http.request.remoteaddr="18.236.180.8:23923" http.request.uri="/v2/library/nginx/blobs/sha256:5430e98eba646ef4a34baff035f6f7483761c873711febd48fbcca38d7890c1e?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=368.70882ms http.response.status=200 http.response.written=1397 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:5430e98eba646ef4a34baff035f6f7483761c873711febd48fbcca38d7890c1e?ns=docker.io HTTP/1.1" 200 1397 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:39.213741826Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=f01a0bd3-1985-4c4b-a888-728486d11566 http.request.method=GET http.request.remoteaddr="18.236.180.8:18627" http.request.uri="/v2/library/nginx/blobs/sha256:03e6a245275128e26fc650e724e3fc4510d81f8111bae35ece70242b0a638215?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=580.83924ms http.response.status=200 http.response.written=895 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:03e6a245275128e26fc650e724e3fc4510d81f8111bae35ece70242b0a638215?ns=docker.io HTTP/1.1" 200 895 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:39.281922202Z" level=info msg="Adding new scheduler entry for library/nginx@sha256:03e6a245275128e26fc650e724e3fc4510d81f8111bae35ece70242b0a638215 with ttl=167h59m59.999983565s" go.version=go1.11.2 instance.id=1470de18-b2bd-4f24-b811-4f48dd848efb service=registry version=v2.7.1 
time="2021-06-23T19:36:39.30334434Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=0a70ac32-9cac-40b9-9810-12ba77fe6607 http.request.method=GET http.request.remoteaddr="18.236.180.8:56809" http.request.uri="/v2/library/nginx/blobs/sha256:b21fed559b9f420d83f8e38ca08d1ac4f15298a3ae02c6de56f364bee2299f78?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=660.705942ms http.response.status=200 http.response.written=602 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:b21fed559b9f420d83f8e38ca08d1ac4f15298a3ae02c6de56f364bee2299f78?ns=docker.io HTTP/1.1" 200 602 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:39.311026941Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=b98667dd-0051-4aef-a026-b7b18b610932 http.request.method=GET http.request.remoteaddr="18.236.180.8:38323" http.request.uri="/v2/library/nginx/blobs/sha256:b82f7f888feb03d38fed4dad68d7265a8b276f1f0c543d549fc6ef30b42c00eb?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=660.928532ms http.response.status=200 http.response.written=667 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:b82f7f888feb03d38fed4dad68d7265a8b276f1f0c543d549fc6ef30b42c00eb?ns=docker.io HTTP/1.1" 200 667 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:39.316264076Z" level=info msg="Adding new scheduler entry for library/nginx@sha256:b21fed559b9f420d83f8e38ca08d1ac4f15298a3ae02c6de56f364bee2299f78 with ttl=167h59m59.999995972s" go.version=go1.11.2 instance.id=1470de18-b2bd-4f24-b811-4f48dd848efb service=registry version=v2.7.1 
time="2021-06-23T19:36:39.395778065Z" level=info msg="Adding new scheduler entry for library/nginx@sha256:b82f7f888feb03d38fed4dad68d7265a8b276f1f0c543d549fc6ef30b42c00eb with ttl=167h59m59.999997262s" go.version=go1.11.2 instance.id=1470de18-b2bd-4f24-b811-4f48dd848efb service=registry version=v2.7.1 
time="2021-06-23T19:36:39.717275884Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=887d2a5f-905d-41ec-9e89-44153fe491d9 http.request.method=GET http.request.remoteaddr="18.236.180.8:25135" http.request.uri="/v2/library/nginx/blobs/sha256:edb81c9bc1f5416a41e5bea21748dc912772fedbd4bd90e5e3ebfe16b453edce?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=1.101527344s http.response.status=200 http.response.written=26580118 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:edb81c9bc1f5416a41e5bea21748dc912772fedbd4bd90e5e3ebfe16b453edce?ns=docker.io HTTP/1.1" 200 26580118 "" "containerd/1.4.6+bottlerocket"
time="2021-06-23T19:36:39.891635154Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=44.234.86.205 http.request.id=0c9a9d08-e289-4d07-ac2d-6c2df642f4d2 http.request.method=GET http.request.remoteaddr="18.236.180.8:24956" http.request.uri="/v2/library/nginx/blobs/sha256:b4d181a07f8025e00e0cb28f1cc14613da2ce26450b80c54aea537fa93cf3bda?ns=docker.io" http.request.useragent="containerd/1.4.6+bottlerocket" http.response.contenttype="application/octet-stream" http.response.duration=1.26689776s http.response.status=200 http.response.written=27145851 
18.236.180.8 - - [23/Jun/2021:19:36:38 +0000] "GET /v2/library/nginx/blobs/sha256:b4d181a07f8025e00e0cb28f1cc14613da2ce26450b80c54aea537fa93cf3bda?ns=docker.io HTTP/1.1" 200 27145851 "" "containerd/1.4.6+bottlerocket"
```

For host-containers and bootstrap-containers:
Started host-container that just runs nginx:
```
bash-5.0# apiclient set host-containers.nginx.source="docker.io/library/nginx:latest"
bash-5.0# apiclient set host-containers.nginx.enabled=true 
```
As soon as the host-container started, I observed traffic through the pull-through cache from my instance:
```
time="2021-07-27T16:24:08.039378021Z" level=info msg="response completed" go.version=go1.11.2 http.request.host=34.222.201.232 http.request.id=e0293cee-59bd-4e5e-987c-f1c9321591b3 http.request.method=HEAD http.request.remoteaddr="34.222.177.43:49756" http.request.uri="/v2/library/nginx/manifests/latest?ns=docker.io" http.request.useragent="containerd/1.4.3+unknown" http.response.contenttype="application/vnd.docker.distribution.manifest.list.v2+json" http.response.duration=729.030479ms http.response.status=200 http.response.written=1862 
34.222.177.43 - - [27/Jul/2021:16:24:07 +0000] "HEAD /v2/library/nginx/manifests/latest?ns=docker.io HTTP/1.1" 200 1862 "" "containerd/1.4.3+unknown"
34.222.177.43 - - [27/Jul/2021:16:24:55 +0000] "HEAD /v2/library/nginx/manifests/latest?ns=docker.io HTTP/1.1" 200 1862 "" "containerd/1.4.3+unknown"
...
```

For bootstrap-containers, I defined a bootstrap container that just runs nginx along with the image registry config in my userdata.
Launched instance and bootstrap container started as expected and I observed corresponding traffic through the pull through cache.

Tested migrations and they work as expected

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
docker, containerd: container image registry mirrors

Adds a new setting `settings.container-registry.mirrors` that lets users set
pull-through caches for container image registries.

tjkirch commented on sources/models/shared-defaults/docker-services.toml:
```diff
@@ -1,3 +1,11 @@
 [services.docker]
 restart-commands = ["/bin/systemctl try-restart docker.service"]
-configuration-files = ["proxy-env"]
+configuration-files = ["docker-daemon-config", "proxy-env"]
```
I think we'll need a migration to change/add the metadata, and a migration to add the new setting, right?

tjkirch commented on README.md:
````diff
@@ -407,6 +407,18 @@ These settings can be changed at any time.
+  ```
+  [settings.image-registry.mirrors]
+  "docker.io" = ["https://<my-docker-hub-mirror-host>", "https://registry-1.docker.io"]
+  "gcr.io" = ["https://<example-gcr-mirror>","https://<example-gcr-mirror-2>"]
````
```suggestion
  "gcr.io" = ["https://<my-gcr-mirror-host>", "https://<example-gcr-mirror-2>"]
```
The suggestion is a nit, but to make the examples consistent so no one thinks there's an important difference.

Question, though - could we use `https://gcr.io` as the second entry in the gcr list, so it's a working entry, like the dockerhub URL in the line above?

etungsten replied:
> Question, though - could we use https://gcr.io as the second entry in the gcr list, so it's a working entry, like the dockerhub URL in the line above?

Yes, we can. But it's also redundant (at least for containerd) since it would try the registry directly if all the mirrors fail. For docker it's not clear so I listed docker hub in the list as well as a fallback.

tjkirch replied:
Ah, ok, that's interesting.  If it's quick to confirm the difference, that'd be nice.  Maybe we could add a comment in the README above this, if it's unexpected?

etungsten replied:
I just tried setting the mirror to some gibberish and ECS was still able to run a task on the node so I guess it fallsback by default. I'll go ahead and fix the example.

tjkirch replied:
It might be worth clarifying (with a comment in README) that it will still fall back to official sources - this isn't a way to insulate yourself from third parties.  That could be useful from the perspectives of debugging and security.

tjkirch commented on README.md:
````diff
@@ -407,6 +407,18 @@ These settings can be changed at any time.
+  "docker.io" = ["https://<my-docker-hub-mirror-host>", "https://registry-1.docker.io"]
+  "gcr.io" = ["https://<example-gcr-mirror>","https://<example-gcr-mirror-2>"]
+  ```
+  Note that for variants that use Docker as its container runtime, Docker only supports pull-through caches for images from Docker Hub (docker.io). Registry mirrors for other registries are ignored in this case.
````
```suggestion
  If you use a Bottlerocket variant that uses Docker as the container runtime, like aws-ecs-1, you should be aware that Docker only supports pull-through caches for images from Docker Hub (docker.io).  Mirrors for other registries are ignored in this case.
```

tjkirch commented on packages/docker-engine/docker-engine.spec:
```diff
@@ -21,7 +21,7 @@ Source0: https://%{repo}/archive/v%{gover}/%{project}-%{gover}.tar.gz
 Source3: docker-sysusers.conf
-Source4: daemon.json
+Source4: daemon-json
 Source5: docker-tmpfiles.conf
```
Won't docker-tmpfiles.conf go away, since it doesn't need to create daemon.json anymore?

tjkirch commented on sources/models/shared-defaults/kubernetes-containerd.toml:
```diff
@@ -1,3 +1,7 @@
+
+# Image registries
+[metadata.settings.image-registry]
+affected-services = ["containerd"]
```
I think having the image-registry line in docker-services makes sense, since that's where the rest of docker is defined, but I'm less sure about this one.  I know it'll work, because every variant either has docker-services or kubernetes-containerd in its defaults, but if we had another type of variant with containerd but not k8s, we'd have to remember to put this there.  The rest of the containerd settings are in defaults.toml, and this file is there to override it with k8s specifics, but I don't think this is k8s-specific.  Should this go in defaults.toml instead?

etungsten replied:
It is k8s-specific since it impacts cri-containerd, which is specific to k8s.

The containerd config file that gets rendered is `containerd-config-toml_k8s`, the generic one being `containerd-config-toml_basic`

tjkirch replied:
Ah, ok.  Do we need a similar block in `containerd-config-toml_basic` then, or is it only possible to use mirrors with the setup that k8s supports?

etungsten replied:
It's a configuration only available to the `cri` containerd plugin. I don't believe containerd itself has a equivalent configuration for registry mirrors.

samuelkarp replied:
With normal usage of containerd (without the cri-containerd plugin), image pull logic (including network transfer) happens in the client program, not in the containerd daemon.  Most client programs don't read containerd's config file.

tjkirch commented on sources/models/src/aws-ecs-1/mod.rs:
```diff
@@ -5,7 +5,7 @@ use std::collections::HashMap;
 use crate::modeled_types::Identifier;
```
I see the new setting in ecs, aws 1.19, and vmware, but shouldn't it also be in aws 1.21, and *-dev?  (Meaning all variants, when you account for symlinks.)

etungsten replied:
All those `mod.rs` are symlinked to one of the `mod.rs` for aws-ecs-1, aws-k8s-1.19 or vmware-k8s-1.21. I didn't add it for the *-dev variants since I thought they didn't need it but on second thought they probably should have it.

tjkirch commented on packages/containerd/containerd-config-toml_k8s:
```diff
@@ -29,3 +29,11 @@ SystemdCgroup = true
 bin_dir = "/opt/cni/bin"
 conf_dir = "/etc/cni/net.d"
+
+{{#if settings.image-registry.mirrors}}
```
nit: tildes at the start of the conditional lines to prevent extra blanks?

etungsten replied:
With tildes at the start of `#if` will remove all white space before the handlebars expression causing the two configuration blocks to merge like this:
```
[plugins."io.containerd.grpc.v1.cri".cni]
bin_dir = "/opt/cni/bin"
conf_dir = "/etc/cni/net.d"
[plugins."io.containerd.grpc.v1.cri".registry.mirrors]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."..."]
        endpoint = [.....]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."..."]
        endpoint = [.....]
```

Which deviates from how the rest of the config file looks (space in between each configuration block). It doesn't change anything functionally though

etungsten commented:
Taking into draft since I forgot to write a new migration for the setting.

etungsten committed:
migrations: migrate `settings.container-registry`, and related models

Adds migrations for the new `container-registry` setting and the new
docker daemon config file and the new config file for host-containers
and bootstrap containers.

tjkirch commented on sources/api/migration/migrations/v1.1.3/image-registry-mirrors/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+/// We also added a new configuration template file for the Docker daemon
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec![
+        "settings.image-registry.mirrors",
```
```suggestion
        "settings.image-registry",
```
To be clearer that we added this whole prefix.

tjkirch commented on README.md:
```diff
@@ -407,6 +407,19 @@ These settings can be changed at any time.
+
+The following setting is optional and allows you to configure image registry mirrors and pull-through caches for your orchestrated containers.
+* `settings.image-registry.mirrors`: A mapping of image registry to a list of image registry URL endpoints.  When pulling an image from a registry, the container runtime will try the endpoints one by one and use the first working one.
+  Even if the default registry is not specified in the list, docker and containerd will still automatically try it at the end with the `https` scheme.
```
```suggestion
  (Docker and containerd will still try the default registry URL if the mirrors fail.)
```
I was a bit confused about the meaning; I think this is clearer?

etungsten commented:
Push below and above addresses @tjkirch's comments.

tjkirch commented:
LGTM assuming migrations test OK.

samuelkarp commented on README.md:
```diff
@@ -407,6 +407,19 @@ These settings can be changed at any time.
   Supported values are `debug`, `info`, `warn`, `error`, and `crit`, and the default is `info`.
 * `settings.ecs.enable-spot-instance-draining`: If the instance receives a spot termination notice, the agent will set the instance's state to `DRAINING`, so the workload can be moved gracefully before the instance is removed. Defaults to `false`.
 
+#### Image registry settings
```
We talk about OS images in this readme too, so it might be worthwhile to call out that this is specific to container images.

```suggestion
#### Container image registry settings
```

samuelkarp commented on README.md:
```diff
@@ -407,6 +407,19 @@ These settings can be changed at any time.
 
+#### Image registry settings
+
+The following setting is optional and allows you to configure image registry mirrors and pull-through caches for your orchestrated containers.
```
Does this only cover orchestrated containers and not host containers?  Should it cover both?

etungsten replied:
Right, this currently only covers orchestrated containers (containers via cri-containerd). 

For host-containers it's not immediately clear to me how I can configure the containerd client to use registry mirrors. I suspect I'd have to modify resolver to accomplish that? I can create a separate issue to track registry mirrors for host-containers.

samuelkarp replied:
Yes, this would be something that would have to be done in `host-ctr`.  I'm not sure if we should have the setting called `settings.container-registry.mirrors` if it only controls orchestrated containers though; how would we change this to handle host containers in the future?

bcressey replied:
It seems like it's just a bug that `host-ctr` and host containers wouldn't support the mirror settings - same as if a component doesn't support the system-wide proxy settings.

bcressey commented on README.md:
````diff
@@ -407,6 +407,19 @@ These settings can be changed at any time.
+#### Container image registry settings
+
+The following setting is optional and allows you to configure image registry mirrors and pull-through caches for your orchestrated containers.
+* `settings.image-registry.mirrors`: A mapping of image registry to a list of image registry URL endpoints.  When pulling an image from a registry, the container runtime will try the endpoints one by one and use the first working one.
+  (Docker and containerd will still try the default registry URL if the mirrors fail.)
+  * Example user data for setting up image registry mirrors:
+  ```
+  [settings.image-registry.mirrors]
+  "docker.io" = ["https://<my-docker-hub-mirror-host>"]
+  "gcr.io" = ["https://<my-gcr-mirror-host>","http://<my-gcr-mirror-host-2>"]
+  ```
````
🚲 🏠 ⚠️

I saw you fixed this in the README , but `container-registry` might be a better term for the setting since "images" are kind of ambiguous.

In the context of a running Bottlerocket it's probably clear enough, but we deal with a lot of different image types - full disk images, filesystem images, etc.

etungsten commented:
Push above rebases onto develop and fixes conflicts

etungsten commented:
Push above addresses @bcressey 's comment.
- "image-registry" -> "container-registry"

tested things and they still work.

etungsten commented:
Missed some dangling references to `image_registries` in models. Push above fixes that.

etungsten commented:
Rebases onto develop and fixes conflicts with migrations.

bcressey commented on packages/containerd/containerd-config-toml_k8s:
```diff
@@ -29,3 +29,11 @@ SystemdCgroup = true
 bin_dir = "/opt/cni/bin"
 conf_dir = "/etc/cni/net.d"
+
+{{~#if settings.container-registry.mirrors}}
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors]
+{{~#each settings.container-registry.mirrors}}
+    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."{{@key}}"]
+        endpoint = [{{join_array ", " this }}]
+{{~/each}}
+{{~/if}}
```
nit: I don't think we need to mimic the Go style of TOML output
```suggestion
{{~#if settings.container-registry.mirrors}}
{{~#each settings.container-registry.mirrors}}
[plugins."io.containerd.grpc.v1.cri".registry.mirrors."{{@key}}"]
endpoint = [{{join_array ", " this }}]
{{~/each}}

{{~/if}}
```

bcressey commented on packages/docker-engine/docker-engine.spec:
```diff
@@ -68,11 +67,9 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_unitdir}/docker.socket
 install -p -m 0644 %{S:3} %{buildroot}%{_cross_sysusersdir}/docker.conf
 
+install -d %{buildroot}%{_cross_templatedir}
 install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/docker
```
Seems like we don't need `%{_cross_factorydir}%{_cross_sysconfdir}/docker` now.

etungsten commented:
Gonna try digging into host-ctr and containerd and see how I can add support for this for host-containers.

etungsten committed:
host-ctr: add support for container image registries

Adds support for configuring container image registrys for
host-containers and bootstrap containers. A lot of the logic is
borrowed from containerd-cri plugin's implementation of image registry
mirrors.

etungsten commented:
Push above addresses comments from @bcressey .

Push above also adds support for image registry mirrors for host-containers and bootstrap-containers.
Fixed up migrations to include new image registry configuration file for host-ctr.

Tested everything and they work as expected.
Tested migrations via upgrade/downgrade and they work as expected.

etungsten commented:
Push above rebases onto develop. Fixes conflicts in go.mod and go.sum

webern commented on packages/kubernetes-1.17/kubelet.service:
```diff
@@ -15,7 +15,8 @@ ExecStartPre=/usr/bin/host-ctr \
     pull-image \
-    --source=${POD_INFRA_CONTAINER_IMAGE}
+    --source=${POD_INFRA_CONTAINER_IMAGE} \
+    --registry-config=/etc/host-containers/host-ctr-registry.config
```
It's OK for this file to be empty? Also, it is sharing the host-ctr's registry config? It makes sense that the setting would be the same for host-ctr and kubelet, but is it weird for kubelet to depend on something in `/etc/host-containers`? Also, is it something that can be set in kubelet's config file?

etungsten replied:
If this file is empty, then the unmarshalled list of mirrors would be empty. So the image would just be pulled from the host as is.

The reason why this is here is because the pause container needs to be pulled and cached before kubelet starts up. We've depended on host-ctr to do that. The registry mirrors would affect that process as well.
It is sharing the same registry config. This isn't something that can be set in kubelet's config file. All container management is done by the containerd runtime. kubelet has no hand in this.

webern commented on sources/api/migration/migrations/v1.1.5/container-registry-mirrors/src/main.rs:
```diff
@@ -0,0 +1,25 @@
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec![
+        "settings.container-registry",
+        "configuration-files.docker-daemon-config",
```
Is this value problematic on k8s?

etungsten replied:
No, there is no action done on forward upgrade. On backwards migration, the value is only removed if it exists. Since k8s won't have this set, it's effectively ignored. There are other similar migrations that does the same thing.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -942,6 +977,47 @@ func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
+// Heavily borrowed from CRI plugin's implementation with the auth related configuration omitted.
+// See https://github.com/containerd/cri/blob/f6026296a3991010429db91e7e677f9c9d4861ab/pkg/server/image_pull.go#L314-L315
+// FIXME Replace this once there's a public containerd client interface that supports registry mirrors
+func registryHosts(registryConfig *RegistryConfig, authorizer docker.Authorizer) docker.RegistryHosts {
```
Can we add some unit tests for this function?

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -879,8 +905,17 @@ func tagImage(ctx context.Context, imageName string, newImageName string, client
 // withDynamicResolver provides an initialized resolver for use with ref.
-func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
-	noOp := func(_ *containerd.Client, _ *containerd.RemoteContext) error { return nil }
+func withDynamicResolver(ctx context.Context, ref string, registryConfig *RegistryConfig) containerd.RemoteOpt {
```
The implementation here has different behavior for private ECR registries compared to all others.  I don't really _expect_ folks to want to use mirrors for ECR, but we should either call that out, update the [resolver](https://github.com/awslabs/amazon-ecr-containerd-resolver/) we're using to support it, or revisit the choice of using that resolver...

etungsten replied:
I'll go ahead and add a comment here to explain that private ECR registries will not use the registry configuration.

I can create an issue in amazon-ecr-containerd-resolver's repository for registry mirror support. I wasn't sure if there is a broader use-case for it.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -942,6 +977,47 @@ func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
+// FIXME Replace this once there's a public containerd client interface that supports registry mirrors
+func registryHosts(registryConfig *RegistryConfig, authorizer docker.Authorizer) docker.RegistryHosts {
+	return func(host string) ([]docker.RegistryHost, error) {
+		var registries []docker.RegistryHost
+		var endpoints []string
```
nit:

```go
var (
	registries []docker.RegistryHost
	endpoints  []string
)
```

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -942,6 +977,47 @@ func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
+		var registries []docker.RegistryHost
+		var endpoints []string
+		_, ok := registryConfig.Mirrors[host]
+		if ok {
+			endpoints = registryConfig.Mirrors[host].Endpoints
+		} else {
+			endpoints = registryConfig.Mirrors["*"].Endpoints
+		}
```
We only support a single wildcard match rather than globbing?

etungsten replied:
Right, this would match up with containerd-cri's behavior. `"*"` here just means for all other hosts that weren't specified in the config, use `"*"`'s backup endpoints.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/registry.go:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	config := RegistryConfig{}
+	toml.Unmarshal(raw, &config)
+	return &config, nil
```
`toml.Unmarshal` returns an error that shouldn't be ignored

```go
return &config, toml.Unmarshal(raw, &config)
```

samuelkarp commented on packages/os/host-ctr-registry-config:
```diff
@@ -0,0 +1,6 @@
+{{~#if settings.container-registry.mirrors}}
+{{~#each settings.container-registry.mirrors}}
+[mirrors."{{@key}}"]
+endpoints = [{{join_array ", " this }}]
+{{~/each}}
+{{~/if}}
```
Is TOML what we're preferring here since it matches containerd's config file format?  Since this isn't exposed we could do something different, and Go has built-in JSON support rather than requiring an external library.

etungsten replied:
I tried rendering it as JSON but didn't work out. See https://github.com/bottlerocket-os/bottlerocket/pull/1629#issuecomment-887994906

samuelkarp replied:
Thanks for testing that; TOML seems like a reasonable choice for now.

etungsten commented:
[Force push](https://github.com/bottlerocket-os/bottlerocket/compare/46e10de5d0e9e0dbbc78880754a9ab3097fef894..3d966d1480e0c1aefed1353b113d74dae7b04c7c) addresses @samuelkarp 's comments.

- Changes the host-ctr-regsitry-config format from TOML to JSON
- Adds unit tests for `registryHosts`
- Adds checks for endpoints without an URL scheme and prefixes them with a default scheme
- Adds comments explaining images from private ECR repositories cannot use registry configuration.

[Force push](https://github.com/bottlerocket-os/bottlerocket/compare/3d966d1480e0c1aefed1353b113d74dae7b04c7c..1ffe8d54929dff4b868a5316b05718f6012a3be9) fixes whitespacing issues with the JSON config template and other small typos.

etungsten commented:
Push above reverts the change to the configuration file format from TOML to JSON.
The handlebars rust crate does not have feature parity with handlebarsjs. It currently does not support `@last`, `@first`, `@index` in `#if` or `#unless` helpers. Which makes it impossible to conditionally render a `,` for items in a list in JSON, so for multiple items in a list, the resulting JSON would be always invalid.

For example, for the following template:
```
{
    {{~#if settings.container-registry.mirrors}}
    "mirrors": {
        {{~#each settings.container-registry.mirrors}}
        "{{@key}}": {
            "endpoints": [{{join_array ", " this }}]
        }{{~#unless @last}}, {{/unless}}
        {{~/each}}
    }
    {{~/if}}
}
```

The `,` always gets rendered even though it shouldn't have.
```
{
    "mirrors": {
        "docker.io": {
            "endpoints": ["http://my-pull-through-cache-host-ip"]
        },
    }
}
```

I tried a bunch of variations of setting the condition, but it just comes down to `#if`/`#unless` not working with `@index`/`@last`/`@first`

We technically can write a new helper that conditionally renders the `,`. But for simplicity's sake, I think we should just stick with TOML. The only downside is that we're taking on an additional `go-toml` dependency.


Tested changes and everything still work as described in the PR testing description.

webern commented:
> We technically can write a new helper that conditionally renders the `,`. 

I already wrote one I though.

arnaldo2792 commented on Release.toml:
```diff
@@ -59,3 +59,7 @@ version = "1.1.4"
     "migrate_v1.1.3_kubelet-cpu-manager.lz4",
 ]
 "(1.1.3, 1.1.4)" = []
+"(1.1.4, 1.1.5)" = [
```
Shouldn't this be 1.2.0? Given that this is a new feature that we will introduce in the next release?

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main_test.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// Test RegistryHosts with invalid endpoints URLs
+func TestBadRegistryHosts(t *testing.T) {
```
nit: since there's just one test case here, I think we can skip the table test format

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -942,6 +979,61 @@ func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
+	}
+}
+
+// Separated out from registryHosts so it is unit-testable. (assert.Equal() does not work on function pointers)
```
nit: this doesn't _have_ to be separated as the test could do something like this:

```go
f := registryHosts(config, authorizer)
result, err := f(host, config, authorizer)
```

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -942,6 +979,61 @@ func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
+		registries []docker.RegistryHost
+		endpoints  []string
+	)
+	_, ok := registryConfig.Mirrors[host]
+	if ok {
```
nit

```suggestion
	if _, ok := registryConfig.Mirrors[host]; ok {
```

samuelkarp reviewed (approved):
Looks good.  A couple nits, but nothing that need to block this.

etungsten commented:
Push above addresses @arnaldo2792 and @samuelkarp 's comments. Will rebase shortly to fix conflicts

etungsten commented:
Push above rebases onto develop and fixes conflicts.

bcressey commented on Release.toml:
```diff
@@ -63,3 +63,7 @@ version = "1.1.4"
+"(1.1.4, 1.2.0)" = [
+    "migrate_v1.2.0_container-registry-mirrors.lz4",
+    "migrate_v1.2.0_container-registry-config-restarts.lz4",
+]
```
Are we planning to fix up migrations in a different commit? If not then these migrations should be with the hostname ones.

bcressey commented on packages/os/host-ctr-registry-config:
```diff
@@ -0,0 +1,6 @@
+{{~#if settings.container-registry.mirrors}}
```
nit: it might be better to call this file `host-ctr.toml` to make room for other potential settings - but I'm also happy to defer that until we need more settings.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -942,6 +979,55 @@ func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
+	}
+}
+
+// registryHosts returns the registry hosts to be used by the resolver. Wraps around _registryHosts.
```
nit: can this comment be updated since `_registryHosts` is gone?

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main_test.go:
```diff
@@ -0,0 +1,161 @@
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			result, err := _registryHosts(tc.host, &tc.config, docker.NewDockerAuthorizer())
```
`_registryHosts` doesn't exist anymore.  ~Looks like we don't yet have automated `go test` runs as part of our build; we can tackle that in a separate PR if you'd prefer but let's make sure the tests that we're checking in run properly.~

Edit: CI is actually failing, so I was wrong about `go test` not being run

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main_test.go:
```diff
@@ -0,0 +1,161 @@
+
+// Test RegistryHosts with invalid endpoints URLs
+func TestBadRegistryHosts(t *testing.T) {
+	t.Run("Messed up mirror endpoint", func(t *testing.T) {
```
nit: I don't think we need the `t.Run` either; this can just be directly in the outer test function

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main_test.go:
```diff
@@ -0,0 +1,161 @@
+// Test RegistryHosts with invalid endpoints URLs
+func TestBadRegistryHosts(t *testing.T) {
+	t.Run("Messed up mirror endpoint", func(t *testing.T) {
+		_, err := _registryHosts("docker.io", &RegistryConfig{
```
Same comment as above `_registryHosts` is now gone.

etungsten commented:
Push above just rebases onto develop and fixes conflicts.

etungsten commented:
Push above addresses comments from @samuelkarp and @bcressey 

- I changed the name of the host-ctr config file from `host-ctr-registry-config` to `host-ctr-toml` to make it more generic.

etungsten commented:
Fixed up Release.toml migrations grouping.
